sqlmesh/core/state_sync/engine_adapter.py

```diff
@@ -58,6 +58,108 @@
 logger = logging.getLogger(__name__)
 
 
+class SnapshotCache(t.MutableMapping[SnapshotId, Snapshot]):
+    def __init__(self, cache: t.Optional[t.Dict[SnapshotId, Snapshot]] = None) -> None:
+        self._cache: t.Dict[SnapshotId, Snapshot] = {}
+        for k, v in (cache or {}).items():
+            self._set_snapshot(k, v)
+
+    def __getitem__(self, key: SnapshotId) -> Snapshot:
+        return self._cache[key]
+
+    def __setitem__(self, key: SnapshotId, value: Snapshot) -> None:
+        existing_snapshot = self.get(key)
+        if existing_snapshot:
+            value = value if value.updated_ts > existing_snapshot.updated_ts else existing_snapshot
+        self._set_snapshot(key, value)
+        self.merge_intervals_by_name_version()
```

Comment:
Why do we merge intervals for **all** snapshots as opposed to only for those that share the version with the target one?

Reply:
This concern is purely performance related correct? There is not a functional issue with the current approach? I agree it is wasteful to merge everything when we know what was updated.